contracts/dmarket.js

```diff
@@ -0,0 +1,1285 @@
+        let qtyTokensToSend = api.BigNumber(sellOrder.price)
+          .multipliedBy(buyOrder.quantity)
+          .toFixed(pairPrecision);
+
+        if (api.BigNumber(qtyTokensToSend).gt(buyOrder.tokensLocked)) {
+          qtyTokensToSend = api.BigNumber(sellOrder.price)
+            .multipliedBy(buyOrder.quantity)
+            .toFixed(pairPrecision, api.BigNumber.ROUND_DOWN);
+        }
+
+        if (api.assert(api.BigNumber(qtyTokensToSend).gt(0)
+          && api.BigNumber(buyOrder.quantity).gt(0), 'the order cannot be filled')) {
+          // transfer the tokens to the buyer
+          await api.transferTokens(account, symbol, buyOrder.quantity, 'user');
+
+          // transfer the tokens to the seller
+          await api.transferTokens(sellOrder.account, pair, qtyTokensToSend, 'user');
+
+          // update the sell order
+          const qtyLeftSellOrder = api.BigNumber(sellOrder.quantity)
+            .minus(buyOrder.quantity)
+            .toFixed(tokenPrecision);
+          const nbTokensToFillOrder = api.BigNumber(sellOrder.price)
+            .multipliedBy(qtyLeftSellOrder)
+            .toFixed(pairPrecision);
+
+          if (api.BigNumber(qtyLeftSellOrder).gt(0)
+            && (api.BigNumber(nbTokensToFillOrder).gte('0.00000001'))) {
+            sellOrder.quantity = qtyLeftSellOrder;
+
+            await api.db.update('sellBook', sellOrder);
+          } else {
+            if (api.BigNumber(qtyLeftSellOrder).gt(0)) {
+              // transfer remaining tokens to seller since the order can no longer be filled
+              await api.transferTokens(sellOrder.account, symbol, qtyLeftSellOrder, 'user');
+            }
+            api.emit('orderClosed', { account: sellOrder.account, type: 'sell', txId: sellOrder.txId });
+            await api.db.remove('sellBook', sellOrder);
+          }
+
+          // unlock remaining tokens, update the quantity to get and remove the buy order
+          const tokensToUnlock = api.BigNumber(buyOrder.tokensLocked)
+            .minus(qtyTokensToSend)
+            .toFixed(pairPrecision);
+
+          if (api.BigNumber(tokensToUnlock).gt(0)) {
+            // transfer any dust tokens remaining to buyer
+            await api.transferTokens(account, pair, tokensToUnlock, 'user');
+          }
+
+          // add the trade to the history
+          await updateTradesHistory('buy', account, sellOrder.account, symbol, pair, pairPrecision, buyOrder.quantity, sellOrder.price, qtyTokensToSend, buyOrder.txId, sellOrder.txId);
+
+          // update the volume
+          volumeTraded = api.BigNumber(volumeTraded).plus(qtyTokensToSend);
+
+          // set quantity to zero to stop the loop
+          buyOrder.quantity = '0';
+          await api.db.remove('buyBook', buyOrder);
+          api.emit('orderClosed', { account: buyOrder.account, type: 'buy', txId: buyOrder.txId });
+        }
+      } else {
+        let qtyTokensToSend = api.BigNumber(sellOrder.price)
```

Comment:
I would like to simplify this loop. This else clause has a lot in common with the previous. Can we consolidate the common logic?

Reply:
I did have a look but it just gets more complicated since there are little key differences. Not sure if possible in a simpler way.